src/Features/Core/Portable/Shared/TestHooks/IAsynchronousOperationListenerProvider.cs

```diff
@@ -0,0 +1,209 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Composition;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Threading.Tasks;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Shared.TestHooks
+{
+    /// <summary>
+    /// Return <see cref="IAsynchronousOperationListener"/> for the given featureName
+    /// 
+    /// We have this abstraction so that we can have isolated listener/waiter in unit tests
+    /// </summary>
+    internal interface IAsynchronousOperationListenerProvider
+    {
+        /// <summary>
+        /// Get <see cref="IAsynchronousOperationListener"/> for given feature.
+        /// same provider will return a singleton listener for same feature
+        /// </summary>
+        IAsynchronousOperationListener GetListener(string featureName);
+    }
+
+    [Export(typeof(IAsynchronousOperationListenerProvider)), Shared]
+    internal class AsynchronousOperationListenerProvider : IAsynchronousOperationListenerProvider
+    {
+        public static readonly IAsynchronousOperationListenerProvider NullProvider = new NullListenerProvider();
+        public static readonly IAsynchronousOperationListener NullListener = new NullOperationListener();
+
+        /// <summary>
+        /// indicate whether asynchronous listener is enabled or not
+        /// </summary>
+        public static bool? s_enabled = null;
+
+        private readonly ConcurrentDictionary<string, IAsynchronousOperationListener> _singletonListeners =
+            new ConcurrentDictionary<string, IAsynchronousOperationListener>(concurrencyLevel: 2, capacity: 20);
+
+        /// <summary>
+        /// indicate whether <see cref="AsynchronousOperationListener.TrackActiveTokens"/> is enabled or not
+        /// </summary>
+        private bool? _trackingBehavior;
+
+        public static void Enable(bool enable)
+        {
+            // right now, made it static so that one can enable it through reflection easy
+            // but we can think of some other way
```

Comment:
I agree with your hatred of static, and agree that it's really damn handy here. :smile: